llvm/lib/Support/Mustache.cpp

```diff
@@ -0,0 +1,489 @@
+SmallString<128> Template::render(Value Data) {
+  BumpPtrAllocator LocalAllocator;
+  return Tree->render(Data, LocalAllocator, Partials, Lambdas, SectionLambdas,
+                      Escapes);
+}
+
+void Template::registerPartial(StringRef Name, StringRef Partial) {
+  Parser P = Parser(Partial, Allocator);
+  ASTNode *PartialTree = P.parse();
+  Partials.insert(std::make_pair(Name, PartialTree));
+}
+
+void Template::registerLambda(StringRef Name, Lambda L) { Lambdas[Name] = L; }
+
+void Template::registerLambda(StringRef Name, SectionLambda L) {
+  SectionLambdas[Name] = L;
+}
+
+void Template::registerEscape(DenseMap<char, StringRef> E) { Escapes = E; }
+
+Template::Template(StringRef TemplateStr) {
+  Parser P = Parser(TemplateStr, Allocator);
+  Tree = P.parse();
+  // the default behaviour is to escape html entities
+  DenseMap<char, StringRef> HtmlEntities = {{'&', "&amp;"},
+                                            {'<', "&lt;"},
+                                            {'>', "&gt;"},
+                                            {'"', "&quot;"},
+                                            {'\'', "&#39;"}};
+  registerEscape(HtmlEntities);
+}
+
+SmallString<128> printJson(Value &Data) {
+
+  SmallString<128> Result;
+  if (Data.getAsNull())
+    return Result;
+  if (auto *Arr = Data.getAsArray())
+    if (Arr->empty())
+      return Result;
+  if (Data.getAsString()) {
+    Result += Data.getAsString()->str();
+    return Result;
+  }
+  if (auto Num = Data.getAsNumber()) {
+    std::ostringstream Oss;
+    Oss << *Num;
+    Result += Oss.str();
+    return Result;
+  }
+  return formatv("{0:2}", Data);
+}
+
+bool isFalsey(Value &V) {
+  return V.getAsNull() || (V.getAsBoolean() && !V.getAsBoolean().value()) ||
+         (V.getAsArray() && V.getAsArray()->empty()) ||
+         (V.getAsObject() && V.getAsObject()->empty());
+}
+
+SmallString<128> ASTNode::render(Value Data, llvm::BumpPtrAllocator &Allocator,
+                                 StringMap<ASTNode *> &Partials,
+                                 StringMap<Lambda> &Lambdas,
+                                 StringMap<SectionLambda> &SectionLambdas,
+                                 DenseMap<char, StringRef> &Escapes) {
+  LocalContext = Data;
+  Value Context = T == Root ? Data : findContext();
+  SmallString<128> Result;
+  switch (T) {
+  case Root: {
+    for (ASTNode *Child : Children)
+      Result += Child->render(Context, Allocator, Partials, Lambdas,
+                              SectionLambdas, Escapes);
+    return Result;
+  }
+  case Text:
+    return Body;
+  case Partial: {
+    if (Partials.find(Accessor[0]) != Partials.end()) {
+      ASTNode *Partial = Partials[Accessor[0]];
+      Result += Partial->render(Data, Allocator, Partials, Lambdas,
+                                SectionLambdas, Escapes);
+      addIndentation(Result, Indentation);
+    }
+    return Result;
+  }
+  case Variable: {
+    if (Lambdas.find(Accessor[0]) != Lambdas.end()) {
```

Comment:
You can avoid a double lookup if you store the result of `find()` in a local and compare it to `end()` in the conditional. Then you don't need the direct access below(`Lambdas[Accessor[0]]`), and can just use the iterator directly.